util/ipgen/renderer.py

```diff
@@ -0,0 +1,263 @@
+            self.ip_template.template_path)
+        template = lookup.get_template(str(template_filepath_rel))
+
+        helper_funcs = {}
+        helper_funcs['unique_vlnv'] = self._tplfunc_unique_vlnv
+
+        # TODO: Introduce namespacing for the template parameters and other
+        # parameters, and/or pass the IpConfig instance to the template after
+        # we have converted more IP blocks to ipgen.
+        tpl_args = {
+            "instance_name": self.ip_config.instance_name,
+            **helper_funcs,
+            **self.get_template_parameter_values() # type: ignore
+        }
+        return template.render(**tpl_args)
+
+    def _render_mako_template_to_file(self, template_filepath: Path,
+                                      outdir_path: Path) -> None:
+        """ Render a file template into a file in outdir_path.
+
+        The name of the output file matches the file name of the template file,
+        without the '.tpl' suffix.
+        """
+
+        assert outdir_path.is_dir()
+
+        outfile_name = self._filename_without_tpl_suffix(template_filepath)
+
+        with open(outdir_path / outfile_name, 'w') as f:
+            f.write(self._render_mako_template_to_str(template_filepath))
+
+    def _filename_without_tpl_suffix(self, filepath: Path) -> str:
+        """ Get the name of the file without a '.tpl' suffix. """
+        assert filepath.suffix == '.tpl'
+        return filepath.stem
+
+
+class IpDescriptionOnlyRenderer(IpTemplateRendererBase):
+    """ Generate the IP description only.
+
+    The IP description is the content of what is typically stored
+    data/ip_name.hjson.
+    """
+    def render(self) -> str:
+        template_path = self.ip_template.template_path
+
+        # Look for a data/ip_name.hjson.tpl template file and render it.
+        hjson_tpl_path = template_path / 'data' / (self.ip_template.name +
+                                                   '.hjson.tpl')
+        if hjson_tpl_path.is_file():
+            return self._render_mako_template_to_str(hjson_tpl_path)
+
+        # Otherwise, if a data/ip_name.hjson file exists, use that.
+        hjson_path = template_path / 'data' / (self.ip_template.name +
+                                               '.hjson')
+        try:
+            with open(hjson_path, 'r') as f:
+                return f.read()
+        except FileNotFoundError:
+            raise FileNotFoundError(
+                "Neither a IP description template at {}, "
+                "nor an IP description at {} exist!".format(
+                    hjson_tpl_path, hjson_path))
+
+
+class IpBlockRenderer(IpTemplateRendererBase):
+    """ Produce a full IP block directory from a template.
+
+    - Copy everything but templates ending with '.tpl' into the destination
+      directory.
+    - Process all Mako templates (files ending in '.tpl') and place them into
+      the output directory. The '.tpl' suffix is stripped off.
```

Comment:
At a quick glance, the first sentence might mean that you do copy templates that don't end with '.tpl'. Maybe it would be better to move the parenthesized bit from the second sentence to the first? Also, we're not actually copying the templates to the output directory. Maybe something like this?
```
- Copy everything but Mako templates (files ending with '.tpl') into the destination directory.
- Process all Mako templates and write the results to the output directory. A template at <PATH>.tpl will write results to <PATH> in the output directory.
```